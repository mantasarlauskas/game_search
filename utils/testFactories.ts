import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
