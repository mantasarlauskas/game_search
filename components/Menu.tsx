import { useEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import MenuItemList from 'components/MenuItemList';
import styles from 'components/Menu.module.scss';
import { Route } from 'utils/routes';
import Link from 'next/link';
import MenuIcon from 'assets/menu.svg';
import { useRouter } from 'next/router';
import { PageTitle } from 'utils/page';
import { useAppContext } from 'pages/_app';

const menuRoutes = [
    Route.HOME,
    Route.DEVELOPERS,
    Route.PUBLISHERS,
    Route.STORES,
    Route.TAGS,
];

function Menu() {
    const [fullscreenVisible, setFullscreenVisible] = useState(false);
    const { genres, platforms, stores } = useAppContext();
    const { asPath } = useRouter();

    useEffect(() => {
        setFullscreenVisible(false);
    }, [asPath]);

    useEffect(() => {
        document.body.style.overflow = fullscreenVisible ? 'hidden' : 'unset';
    }, [fullscreenVisible]);

    const menu = (
        <>
            {menuRoutes.map((route) => (
                <Link key={route} href={route}>
                    <div className={styles.item}>{PageTitle[route]}</div>
                </Link>
            ))}
            <div className={styles.list}>
                <MenuItemList items={genres} route={Route.GENRES} />
            </div>
            <div className={styles.list}>
                <MenuItemList
                    items={platforms}
                    expandedCount={8}
                    route={Route.PLATFORMS}
                />
            </div>
            <div className={styles.list}>
                <MenuItemList items={stores} route={Route.STORES} />
            </div>
        </>
    );

    if (fullscreenVisible) {
        return createPortal(
            <div className={styles.portal}>
                <MenuIcon
                    onClick={() => setFullscreenVisible(false)}
                    className={styles.icon}
                />
                {menu}
            </div>,
            document.body
        );
    }

    return (
        <div className={styles.root}>
            <MenuIcon
                onClick={() => setFullscreenVisible(true)}
                className={styles.icon}
            />
            <div className={styles.content}>{menu}</div>
        </div>
    );
}

export default Menu;
