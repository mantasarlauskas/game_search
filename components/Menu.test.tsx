import { render, screen } from '@testing-library/react';
import Menu from 'components/Menu';
import { getCategories } from 'testing/testFactories';

const mockedCategories = getCategories();
jest.mock('pages/_app', () => ({
    useAppContext() {
        return {
            genres: [mockedCategories[0]],
            platforms: [mockedCategories[1]],
            stores: [],
        };
    },
}));

jest.mock('next/router', () => ({
    useRouter() {
        return {
            asPath: '',
        };
    },
}));

describe('<Menu />', () => {
    it('matches snapshot', () => {
        const { container } = render(<Menu />);
        expect(container).toMatchSnapshot();
    });

    it('shows menu items', () => {
        render(<Menu />);
        expect(screen.getByText('Feral Interactive')).toBeInTheDocument();
        expect(screen.getByAltText('Feral Interactive')).toBeInTheDocument();
        expect(screen.getByText('Ubisoft')).toBeInTheDocument();
        expect(screen.getByAltText('Ubisoft')).toBeInTheDocument();
    });
});
