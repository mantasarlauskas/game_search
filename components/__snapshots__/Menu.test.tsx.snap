// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <span />
    <div
      class="content"
    >
      <div>
        <div
          class="item"
        >
          Home
        </div>
      </div>
      <div>
        <div
          class="item"
        >
          Developers
        </div>
      </div>
      <div>
        <div
          class="item"
        >
          Publishers
        </div>
      </div>
      <div>
        <div
          class="item"
        >
          Stores
        </div>
      </div>
      <div>
        <div
          class="item"
        >
          Tags
        </div>
      </div>
      <div
        class="list"
      >
        <div
          class="root"
        >
          <div>
            <div
              class="title"
            >
              Genres
            </div>
          </div>
          <div>
            <div
              class="item"
            >
              <img
                alt="Feral Interactive"
                class="background"
                src="https://media.rawg.io/media/crop/600/400/games/157/15742f2f67eacff546738e1ab5c19d20.jpg"
              />
              <div
                class="name"
              >
                Feral Interactive
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="list"
      >
        <div
          class="root"
        >
          <div>
            <div
              class="title"
            >
              Platforms
            </div>
          </div>
          <div>
            <div
              class="item"
            >
              <img
                alt="Ubisoft"
                class="background"
                src="https://media.rawg.io/media/crop/600/400/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg"
              />
              <div
                class="name"
              >
                Ubisoft
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="list"
      >
        <div
          class="root"
        >
          <div>
            <div
              class="title"
            >
              Stores
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
